File: Task1008/Sources/Task1008/validation.swift
import Foundation
#if DEVELOPING
import TestData

let data1 = TestData(name: "Data 1", text: """
4 3
1 3 6
0 1 11
1 2 7
""", expected: """
0
11
18
17
""")

let data2 = TestData(name: "Data 2", text: """
5 10
3 1 46
2 1 48
4 2 21
0 1 26
4 1 50
4 0 87
1 0 57
1 3 42
1 2 52
0 4 10
""", expected: """
0
26
31
68
10
""")

let data3 = TestData(name: "Data 3", text: """
""", expected: """
""")

let data4 = TestData(name: "Data 4", text: """
""", expected: """
""")

let data5 = TestData(name: "Data 5", text: """
""", expected: """
""")

let validationList = [
    data1,
    data2,
//    data3,
//    data4,
//    data5,
]

// ----------------------------------------------------------
func print(_ items: Any...) {
    TestData.write(items)
}

func validate(_ main: (() -> String) -> Void) {
    validationList.forEach { $0.execute(main) }
}
// ----------------------------------------------------------

#endif

File: template-kotlin/src/Validation.kt
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.PrintStream

fun main() {
    val data1 = TestData("Data 1", """
        input
        1
        2 3
    """.trimIndent(), """
        input 1 2 3
    """.trimIndent())

    val data2 = TestData("Data 2", """
    """.trimIndent(), """
    """.trimIndent())

    val data3 = TestData("Data 3", """
    """.trimIndent(), """
    """.trimIndent())

    val data4 = TestData("Data 4", """
    """.trimIndent(), """
    """.trimIndent())

    val data5 = TestData("Data 5", """
    """.trimIndent(), """
    """.trimIndent())

    val validationList = listOf(
        data1,
//        data2,
//        data3,
//        data4,
//        data5,
    )
    
    println()
    validationList.forEach { it.execute(::solve) }
    println("--------------------------------")
    println()
}

class TestData(
    private val name: String,
    private val data: String,
    private val expected: String? = null,
) {
    fun execute(block: () -> Unit) {
        println("--- [$name] -------------------")
        val originalIn = System.`in`
        val originalOut = System.out
        val actual = ByteArrayInputStream(data.toByteArray()).use { inputStream ->
            ByteArrayOutputStream().use { outputStream ->
                System.setIn(inputStream)
                try {
                    System.setOut(PrintStream(outputStream))
                    try {
                        block()
                        outputStream.toString()
                    } finally {
                        System.setOut(originalOut)
                    }
                } finally {
                    System.setIn(originalIn)
                }
            }
        }
        print(actual)
        if (actual[actual.length - 1] != '\n') {
            println()
        }

        validate(actual)?.let {
            if (it) {
                println("✅ OK")
            } else {
                println("❌ NG")
            }
        }
    }

    private data class LineReader(val data: String) {
        private val lines = data.split("\n").map { it.trim() }
        private var nextLine = 0

        fun readln(): String {
            return lines[nextLine++]
        }

        fun readlnOrNull(): String? {
            if (nextLine < lines.size) {
                return lines[nextLine++]
            } else {
                return null
            }
        }
    }

    private fun validate(actual: String): Boolean? {
        if (expected == null) {
            return null
        } else {
            val actualReader = LineReader(actual)
            val expectedReader = LineReader(expected)
            var expectedLine = expectedReader.readlnOrNull()
            while (expectedLine != null) {
                val actualLine = actualReader.readlnOrNull()
                if (expectedLine != actualLine) {
                    return false
                }
                expectedLine = expectedReader.readlnOrNull()
            }
            return true
        }
    }
}
